web/packages/teleterm/src/ui/DocumentTerminal/Terminal/ctrl.ts

```diff
@@ -79,12 +84,85 @@ export default class TtyTerminal {
       },
     });
 
+    this.term.onSelectionChange(() => {
+      if (
+        this.configService.get('terminal.copyOnSelect').value &&
+        this.term.hasSelection()
+      ) {
+        void this.copy();
+      }
+    });
+
     this.term.loadAddon(this.fitAddon);
 
     this.registerResizeHandler();
 
     this.term.open(this.el);
 
+    this.term.attachCustomKeyEventHandler(e => {
+      const action = this.keyboardShortcutsService.getShortcutAction(e);
+      const isKeyDown = e.type === 'keydown';
+      if (action === 'terminalCopy' && isKeyDown) {
+        void this.copy();
+        // Do not invoke a copy action from the menu.
+        e.preventDefault();
+        // Event handled, do not process it in xterm.
+        return false;
+      }
+      if (action === 'terminalPaste' && isKeyDown) {
+        void this.paste();
+        // Do not invoke a copy action from the menu.
+        e.preventDefault();
+        // Event handled, do not process it in xterm.
+        return false;
+      }
+
+      return true;
+    });
+
+    this.term.element.addEventListener('contextmenu', e => {
+      // We always call preventDefault because:
+      // 1. When `terminalRightClick` is not `menu`, we don't want to show it.
+      // 2. When `terminalRightClick` is `menu`, opening two menus at
+      //  the same time on Linux causes flickering.
+      e.preventDefault();
+
+      if (this.configService.get('terminal.rightClick').value === 'menu') {
```

Comment:
Initially, I wanted to pass `terminalRightClick` as another property in options, but maybe it makes more sense to pass the entire service? It is easy to mock in tests anyway.

Another thing is that this allows us to dynamically read the config, so when one day we make the config panel, new values will be applied without an app restart or opening a new terminal.